src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs

```diff
@@ -253,9 +254,161 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadDefinitionId workloadId)
+                    }
+                    else
+                    {
+                        incompleteCandidates.Add(candidate);
+                    }
+                }
+            }
+
+            //find all valid complete permutations by recursively exploring possible branches from a root
+            void FindCandidates (WorkloadSuggestionCandidate root, List<WorkloadSuggestionCandidate> branches)
+            {
+                foreach (var branch in branches)
+                {
+                    //skip branches identical to ones that have already already been taken
+                    //there's probably a more efficient way to do this but this is easy to reason about
+                    if (root.Workloads.IsSupersetOf(branch.Workloads))
+                    {
+                        continue;
+                    }
+
+                    //skip branches that don't reduce the number of missing packs
+                    //the branch may be a more optimal solution, but this will be handled elsewhere in the permutation space where it is treated as a root
+                    if (!root.StillMissingPacks.Overlaps(branch.Packs))
+                    {
+                        continue;
+                    }
+
+                    //construct the new condidate by combining the root and the branch
+                    var combinedIds = new HashSet<WorkloadDefinitionId>(root.Workloads);
+                    combinedIds.UnionWith(branch.Workloads);
+                    var combinedPacks = new HashSet<WorkloadPackId>(root.Packs);
+                    combinedPacks.UnionWith(branch.Packs);
+                    var stillMissing = new HashSet<WorkloadPackId>(root.StillMissingPacks);
+                    stillMissing.ExceptWith(branch.Packs);
+                    var candidate = new WorkloadSuggestionCandidate(combinedIds, combinedPacks, stillMissing);
+
+                    //if the candidate contains all the requested packs, it's complete. else, recurse to try adding more branches to it.
+                    if (candidate.IsComplete)
+                    {
+                        completeCandidates.Add(candidate);
+                    }
+                    else
+                    {
+                        FindCandidates(candidate, branches);
+                    }
+                }
+            }
+
+            foreach (var c in incompleteCandidates)
+            {
+                FindCandidates(c, incompleteCandidates);
+            }
+
+            // minimize number of unnecessary packs installed by the suggestion, then minimize number of workloads ids listed in the suggestion
+            // eventually a single pass to find the "best" would be more efficient but this this more readable and easier to tweak
+            var scoredList = completeCandidates.Select(
+                c =>
+                {
+                    var installedCount = c.Packs.Count(p => installedPacks.Contains(p));
```

Comment:
no obvious test for the scores. This part sounds super complex. If a later change messed up the score logic, a test need to fail.

And it is complicated enough that I cannot say it really works by just reading it.